File: apps/nodejs/environment.js
/**
 * Environmental effects simulation.
 *
 * Models random environmental perturbations that affect the evolution
 * of the universe from quantum to biological scales:
 * - Temperature gradients and thermal fluctuations
 * - Radiation (UV, cosmic rays, stellar winds)
 * - Geological events (volcanic, tectonic)
 * - Atmospheric composition changes
 * - Day/night cycles and seasonal variations
 */

import {
    T_PLANCK, T_CMB, T_EARTH_SURFACE, T_STELLAR_CORE,
    UV_MUTATION_RATE, COSMIC_RAY_MUTATION_RATE,
    RADIATION_DAMAGE_THRESHOLD, THERMAL_FLUCTUATION, K_B,
    PI,
} from './constants.js';


// === Helper: random gaussian ===

function gaussRandom(mean = 0, stddev = 1) {
    const u1 = Math.random();
    const u2 = Math.random();
    const z = Math.sqrt(-2.0 * Math.log(u1 || 1e-20)) * Math.cos(2.0 * PI * u2);
    return mean + stddev * z;
}

function expoVariate(lambda) {
    return -Math.log(Math.random() || 1e-20) / lambda;
}


// === EnvironmentalEvent ===

export class EnvironmentalEvent {
    /** A discrete environmental event. */
    constructor(eventType, intensity, duration, position = [0.0, 0.0, 0.0], tickOccurred = 0) {
        this.eventType = eventType;
        this.intensity = intensity;
        this.duration = duration;  // ticks
        this.position = [...position];
        this.tickOccurred = tickOccurred;
    }

    toCompact() {
        return `Ev:${this.eventType}(i=${this.intensity.toFixed(2)},d=${this.duration})`;
    }
}


// === Environment ===

export class Environment {
    /** The physical environment that affects all simulation levels. */
    constructor(initialTemperature = T_PLANCK) {
        this.temperature = initialTemperature;
        this.uvIntensity = 0.0;
        this.cosmicRayFlux = 0.0;
        this.stellarWind = 0.0;
        this.atmosphericDensity = 0.0;
        this.waterAvailability = 0.0;
        this.dayNightCycle = 0.0;  // 0-1, 0=midnight, 0.5=noon
        this.season = 0.0;  // 0-1, 0=winter, 0.5=summer
        this.events = [];
        this.eventHistory = [];
        this.tick = 0;
    }

    /** Update environment based on cosmic epoch. */
    update(epoch) {
        this.tick++;

        // Temperature evolution (logarithmic cooling)
        if (epoch < 1000) {
            // Early universe: rapid cooling
            this.temperature = T_PLANCK * Math.exp(-epoch / 200);
        } else if (epoch < 50000) {
            // Pre-recombination
            this.temperature = Math.max(
                T_CMB, T_PLANCK * Math.exp(-epoch / 200)
            );
        } else if (epoch < 200000) {
            // Post-recombination to star formation
            this.temperature = T_CMB + gaussRandom(0, 0.5);
        } else {
            // Planet era
            this.temperature = T_EARTH_SURFACE + gaussRandom(0, 5);
            // Day/night
            this.dayNightCycle = (this.tick % 100) / 100.0;
            const tempMod = 10 * Math.sin(this.dayNightCycle * 2 * PI);
            this.temperature += tempMod;
            // Seasons
            this.season = (this.tick % 1000) / 1000.0;
            const seasonMod = 15 * Math.sin(this.season * 2 * PI);
            this.temperature += seasonMod;
        }

        // UV intensity (appears with stars)
        if (epoch > 100000) {
            const baseUv = 1.0;
            if (this.dayNightCycle > 0.25 && this.dayNightCycle < 0.75) {
                this.uvIntensity = baseUv * Math.sin(
                    (this.dayNightCycle - 0.25) * 2 * PI
                );
            } else {
                this.uvIntensity = 0.0;
            }
        } else {
            this.uvIntensity = 0.0;
        }

        // Cosmic ray flux
        if (epoch > 10000) {
            this.cosmicRayFlux = 0.1 + expoVariate(10.0);
        } else {
            this.cosmicRayFlux = 1.0;  // High in early universe
        }

        // Atmospheric density (grows with planet formation)
        if (epoch > 210000) {
            this.atmosphericDensity = Math.min(1.0, (epoch - 210000) / 50000);
            // Atmosphere shields from UV
            this.uvIntensity *= (1 - this.atmosphericDensity * 0.7);
            this.cosmicRayFlux *= (1 - this.atmosphericDensity * 0.5);
        }

        // Water availability
        if (epoch > 220000) {
            this.waterAvailability = Math.min(1.0, (epoch - 220000) / 30000);
        }

        // Random events
        this._generateEvents(epoch);

        // Process active events
        this._processEvents();
    }

    _generateEvents(epoch) {
        // Volcanic activity
        if (epoch > 210000 && Math.random() < 0.005) {
            this.events.push(new EnvironmentalEvent(
                "volcanic",
                Math.random() * 2.5 + 0.5,  // uniform(0.5, 3.0)
                Math.floor(Math.random() * 91) + 10,  // randint(10, 100)
                [0.0, 0.0, 0.0],
                this.tick,
            ));
        }

        // Asteroid impact
        if (Math.random() < 0.0001) {
            this.events.push(new EnvironmentalEvent(
                "asteroid",
                Math.random() * 9.0 + 1.0,  // uniform(1.0, 10.0)
                Math.floor(Math.random() * 451) + 50,  // randint(50, 500)
                [0.0, 0.0, 0.0],
                this.tick,
            ));
        }

        // Solar flare
        if (epoch > 100000 && Math.random() < 0.01) {
            this.events.push(new EnvironmentalEvent(
                "solar_flare",
                Math.random() * 1.9 + 0.1,  // uniform(0.1, 2.0)
                Math.floor(Math.random() * 16) + 5,  // randint(5, 20)
                [0.0, 0.0, 0.0],
                this.tick,
            ));
        }

        // Ice age
        if (epoch > 250000 && Math.random() < 0.001) {
            this.events.push(new EnvironmentalEvent(
                "ice_age",
                Math.random() * 1.0 + 0.5,  // uniform(0.5, 1.5)
                Math.floor(Math.random() * 1501) + 500,  // randint(500, 2000)
                [0.0, 0.0, 0.0],
                this.tick,
            ));
        }
    }

    _processEvents() {
        const active = [];
        for (const event of this.events) {
            const elapsed = this.tick - event.tickOccurred;
            if (elapsed < event.duration) {
                active.push(event);
                this._applyEvent(event);
            } else {
                this.eventHistory.push(event);
            }
        }
        this.events = active;
    }

    _applyEvent(event) {
        if (event.eventType === "volcanic") {
            this.temperature += event.intensity * 2;
            this.atmosphericDensity = Math.min(
                1.0, this.atmosphericDensity + 0.01
            );
            this.uvIntensity *= 0.9;  // Ash blocks UV

        } else if (event.eventType === "asteroid") {
            this.temperature -= event.intensity * 5;  // Nuclear winter
            this.cosmicRayFlux += event.intensity;
            this.uvIntensity *= 0.5;

        } else if (event.eventType === "solar_flare") {
            this.uvIntensity += event.intensity;
            this.cosmicRayFlux += event.intensity * 0.5;

        } else if (event.eventType === "ice_age") {
            this.temperature -= event.intensity * 20;
            this.waterAvailability *= 0.8;
        }
    }

    /** Total radiation dose from all sources. */
    getRadiationDose() {
        return this.uvIntensity + this.cosmicRayFlux + this.stellarWind;
    }

    /** Check if conditions support life. */
    isHabitable() {
        return (
            this.temperature > 200
            && this.temperature < 400
            && this.waterAvailability > 0.1
            && this.getRadiationDose() < RADIATION_DAMAGE_THRESHOLD
        );
    }

    /**
     * Available thermal energy for biological processes.
     * Scaled to provide meaningful energy in habitable range.
     * At ~300K, returns ~30 units (enough to sustain cell metabolism).
     */
    thermalEnergy() {
        if (this.temperature < 100 || this.temperature > 500) {
            return 0.1;
        }
        return Math.max(0.1, this.temperature * 0.1);
    }

    toCompact() {
        return (
            `Env[T=${this.temperature.toFixed(1)} `
            + `UV=${this.uvIntensity.toFixed(3)} `
            + `CR=${this.cosmicRayFlux.toFixed(3)} `
            + `atm=${this.atmosphericDensity.toFixed(2)} `
            + `H2O=${this.waterAvailability.toFixed(2)} `
            + `hab=${this.isHabitable() ? "Y" : "N"} `
            + `ev=${this.events.length}]`
        );
    }
}

File: simulator/environment.py
"""Environmental effects simulation.

Models random environmental perturbations that affect the evolution
of the universe from quantum to biological scales:
- Temperature gradients and thermal fluctuations
- Radiation (UV, cosmic rays, stellar winds)
- Geological events (volcanic, tectonic)
- Atmospheric composition changes
- Day/night cycles and seasonal variations
"""
import math
import random
from dataclasses import dataclass, field
from typing import Optional

from simulator.constants import (
    T_PLANCK, T_CMB, T_EARTH_SURFACE, T_STELLAR_CORE,
    UV_MUTATION_RATE, COSMIC_RAY_MUTATION_RATE,
    RADIATION_DAMAGE_THRESHOLD, THERMAL_FLUCTUATION, K_B,
)


@dataclass
class EnvironmentalEvent:
    """A discrete environmental event."""
    event_type: str
    intensity: float
    duration: int  # ticks
    position: list = field(default_factory=lambda: [0.0, 0.0, 0.0])
    tick_occurred: int = 0

    def to_compact(self) -> str:
        return f"Ev:{self.event_type}(i={self.intensity:.2f},d={self.duration})"


class Environment:
    """The physical environment that affects all simulation levels."""

    def __init__(self, initial_temperature: float = T_PLANCK):
        self.temperature = initial_temperature
        self.uv_intensity = 0.0
        self.cosmic_ray_flux = 0.0
        self.stellar_wind = 0.0
        self.atmospheric_density = 0.0
        self.water_availability = 0.0
        self.day_night_cycle = 0.0  # 0-1, 0=midnight, 0.5=noon
        self.season = 0.0  # 0-1, 0=winter, 0.5=summer
        self.events: list[EnvironmentalEvent] = []
        self.event_history: list[EnvironmentalEvent] = []
        self.tick = 0

    def update(self, epoch: int):
        """Update environment based on cosmic epoch."""
        self.tick += 1

        # Temperature evolution (logarithmic cooling)
        if epoch < 1000:
            # Early universe: rapid cooling
            self.temperature = T_PLANCK * math.exp(-epoch / 200)
        elif epoch < 50000:
            # Pre-recombination
            self.temperature = max(
                T_CMB, T_PLANCK * math.exp(-epoch / 200)
            )
        elif epoch < 200000:
            # Post-recombination to star formation
            self.temperature = T_CMB + random.gauss(0, 0.5)
        else:
            # Planet era
            self.temperature = T_EARTH_SURFACE + random.gauss(0, 5)
            # Day/night
            self.day_night_cycle = (self.tick % 100) / 100.0
            temp_mod = 10 * math.sin(self.day_night_cycle * 2 * math.pi)
            self.temperature += temp_mod
            # Seasons
            self.season = (self.tick % 1000) / 1000.0
            season_mod = 15 * math.sin(self.season * 2 * math.pi)
            self.temperature += season_mod

        # UV intensity (appears with stars)
        if epoch > 100000:
            base_uv = 1.0
            # Day/night modulation
            if self.day_night_cycle > 0.25 and self.day_night_cycle < 0.75:
                self.uv_intensity = base_uv * math.sin(
                    (self.day_night_cycle - 0.25) * 2 * math.pi
                )
            else:
                self.uv_intensity = 0.0
        else:
            self.uv_intensity = 0.0

        # Cosmic ray flux
        if epoch > 10000:
            self.cosmic_ray_flux = 0.1 + random.expovariate(10.0)
        else:
            self.cosmic_ray_flux = 1.0  # High in early universe

        # Atmospheric density (grows with planet formation)
        if epoch > 210000:
            self.atmospheric_density = min(1.0, (epoch - 210000) / 50000)
            # Atmosphere shields from UV
            self.uv_intensity *= (1 - self.atmospheric_density * 0.7)
            self.cosmic_ray_flux *= (1 - self.atmospheric_density * 0.5)

        # Water availability
        if epoch > 220000:
            self.water_availability = min(
                1.0, (epoch - 220000) / 30000
            )

        # Random events
        self._generate_events(epoch)

        # Process active events
        self._process_events()

    def _generate_events(self, epoch: int):
        """Generate random environmental events."""
        # Volcanic activity
        if epoch > 210000 and random.random() < 0.005:
            self.events.append(EnvironmentalEvent(
                event_type="volcanic",
                intensity=random.uniform(0.5, 3.0),
                duration=random.randint(10, 100),
                tick_occurred=self.tick,
            ))

        # Asteroid impact
        if random.random() < 0.0001:
            self.events.append(EnvironmentalEvent(
                event_type="asteroid",
                intensity=random.uniform(1.0, 10.0),
                duration=random.randint(50, 500),
                tick_occurred=self.tick,
            ))

        # Solar flare
        if epoch > 100000 and random.random() < 0.01:
            self.events.append(EnvironmentalEvent(
                event_type="solar_flare",
                intensity=random.uniform(0.1, 2.0),
                duration=random.randint(5, 20),
                tick_occurred=self.tick,
            ))

        # Ice age
        if epoch > 250000 and random.random() < 0.001:
            self.events.append(EnvironmentalEvent(
                event_type="ice_age",
                intensity=random.uniform(0.5, 1.5),
                duration=random.randint(500, 2000),
                tick_occurred=self.tick,
            ))

    def _process_events(self):
        """Process active environmental events."""
        active = []
        for event in self.events:
            elapsed = self.tick - event.tick_occurred
            if elapsed < event.duration:
                active.append(event)
                self._apply_event(event)
            else:
                self.event_history.append(event)

        self.events = active

    def _apply_event(self, event: EnvironmentalEvent):
        """Apply an environmental event's effects."""
        if event.event_type == "volcanic":
            self.temperature += event.intensity * 2
            self.atmospheric_density = min(
                1.0, self.atmospheric_density + 0.01
            )
            self.uv_intensity *= 0.9  # Ash blocks UV

        elif event.event_type == "asteroid":
            self.temperature -= event.intensity * 5  # Nuclear winter
            self.cosmic_ray_flux += event.intensity
            self.uv_intensity *= 0.5

        elif event.event_type == "solar_flare":
            self.uv_intensity += event.intensity
            self.cosmic_ray_flux += event.intensity * 0.5

        elif event.event_type == "ice_age":
            self.temperature -= event.intensity * 20
            self.water_availability *= 0.8

    def get_radiation_dose(self) -> float:
        """Total radiation dose from all sources."""
        return self.uv_intensity + self.cosmic_ray_flux + self.stellar_wind

    def is_habitable(self) -> bool:
        """Check if conditions support life."""
        return (
            200 < self.temperature < 400
            and self.water_availability > 0.1
            and self.get_radiation_dose() < RADIATION_DAMAGE_THRESHOLD
        )

    def thermal_energy(self) -> float:
        """Available thermal energy for biological processes.

        Scaled to provide meaningful energy in habitable range.
        At ~300K, returns ~30 units (enough to sustain cell metabolism).
        """
        if self.temperature < 100 or self.temperature > 500:
            return 0.1
        return max(0.1, self.temperature * 0.1)

    def to_compact(self) -> str:
        return (f"Env[T={self.temperature:.1f} "
                f"UV={self.uv_intensity:.3f} "
                f"CR={self.cosmic_ray_flux:.3f} "
                f"atm={self.atmospheric_density:.2f} "
                f"H2O={self.water_availability:.2f} "
                f"hab={'Y' if self.is_habitable() else 'N'} "
                f"ev={len(self.events)}]")
